
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>ACMG guideline &#8212; CharGer  documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Development Guide" href="development.html" />
    <link rel="prev" title="Command line interface" href="cli.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="acmg-guideline">
<h1>ACMG guideline<a class="headerlink" href="#acmg-guideline" title="Permalink to this headline">¶</a></h1>
<p>The table and the module definitions are adapted from the original ACMG
guideline publication<a class="footnote-reference brackets" href="#acmg2015" id="id1">1</a>.</p>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<table class="docutils align-default">
<colgroup>
<col style="width: 24%" />
<col style="width: 11%" />
<col style="width: 16%" />
<col style="width: 16%" />
<col style="width: 13%" />
<col style="width: 11%" />
<col style="width: 11%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ACMG modules</p></th>
<th class="head"><p>Strong</p></th>
<th class="head"><p>Supporting</p></th>
<th class="head"><p>Supporting</p></th>
<th class="head"><p>Moderate</p></th>
<th class="head"><p>Strong</p></th>
<th class="head"><p>Very
Strong</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Population data</p></td>
<td><p><a class="reference internal" href="#term-ba1"><span class="xref std std-term">BA1</span></a>,
<a class="reference internal" href="#term-bs1"><span class="xref std std-term">BS1</span></a>,
<a class="reference internal" href="#term-bs2"><span class="xref std std-term">BS2</span></a></p></td>
<td></td>
<td></td>
<td><p><a class="reference internal" href="#term-pm2"><span class="xref std std-term">PM2</span></a></p></td>
<td><p><a class="reference internal" href="#term-ps4"><span class="xref std std-term">PS4</span></a></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>Computational
and
predictive
data</p></td>
<td></td>
<td><p><a class="reference internal" href="#term-bp1"><span class="xref std std-term">BP1</span></a>,
<a class="reference internal" href="#term-bp4"><span class="xref std std-term">BP4</span></a>,
<a class="reference internal" href="#term-bp4"><span class="xref std std-term">BP4</span></a>,
<a class="reference internal" href="#term-bp7"><span class="xref std std-term">BP7</span></a>,</p></td>
<td><p><a class="reference internal" href="#term-pp3"><span class="xref std std-term">PP3</span></a></p></td>
<td><p><a class="reference internal" href="#term-pm4"><span class="xref std std-term">PM4</span></a>,
<a class="reference internal" href="#term-pm5"><span class="xref std std-term">PM5</span></a></p></td>
<td><p><a class="reference internal" href="#term-ps1"><span class="xref std std-term">PS1</span></a></p></td>
<td><p><a class="reference internal" href="#term-pvs1"><span class="xref std std-term">PVS1</span></a></p></td>
</tr>
<tr class="row-even"><td><p>Functional data</p></td>
<td><p><a class="reference internal" href="#term-bs3"><span class="xref std std-term">BS3</span></a></p></td>
<td></td>
<td><p><a class="reference internal" href="#term-pp2"><span class="xref std std-term">PP2</span></a></p></td>
<td><p><a class="reference internal" href="#term-pm1"><span class="xref std std-term">PM1</span></a></p></td>
<td><p><a class="reference internal" href="#term-ps3"><span class="xref std std-term">PS3</span></a></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>Segregation data</p></td>
<td><p><a class="reference internal" href="#term-bs4"><span class="xref std std-term">BS4</span></a></p></td>
<td></td>
<td colspan="3"><p><a class="reference internal" href="#term-pp1"><span class="xref std std-term">PP1</span></a></p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>De novo data</p></td>
<td></td>
<td></td>
<td></td>
<td><p><a class="reference internal" href="#term-pm6"><span class="xref std std-term">PM6</span></a></p></td>
<td><p><a class="reference internal" href="#term-ps2"><span class="xref std std-term">PS2</span></a></p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>Allele data</p></td>
<td></td>
<td><p><a class="reference internal" href="#term-bp2"><span class="xref std std-term">BP2</span></a></p></td>
<td></td>
<td><p><a class="reference internal" href="#term-pm3"><span class="xref std std-term">PM3</span></a></p></td>
<td></td>
<td></td>
</tr>
<tr class="row-even"><td><p>Others</p></td>
<td></td>
<td><p><a class="reference internal" href="#term-bp5"><span class="xref std std-term">BP5</span></a>,
<a class="reference internal" href="#term-bp6"><span class="xref std std-term">BP6</span></a></p></td>
<td><p><a class="reference internal" href="#term-pp5"><span class="xref std std-term">PP5</span></a>,
<a class="reference internal" href="#term-pp4"><span class="xref std std-term">PP4</span></a></p></td>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="module-definitions-and-criteria">
<h2>Module definitions and criteria<a class="headerlink" href="#module-definitions-and-criteria" title="Permalink to this headline">¶</a></h2>
<dl class="glossary simple">
<dt id="term-pvs1">PVS1</dt><dd><p>Null variant (nonsense, frameshift, canonical ±1 or 2 splice sites,
initiation codon, single or multi-exon deletion) in a gene where LOF
is a known mechanism of disease.</p>
</dd>
<dt id="term-ps1">PS1</dt><dd><p>Same amino acid change as a previously established pathogenic variant
regardless of nucleotide change.</p>
</dd>
<dt id="term-ps2">PS2</dt><dd><p><em>De novo</em> (both maternity and paternity confirmed) in a patient with
the disease and no family history.</p>
</dd>
<dt id="term-ps3">PS3</dt><dd><p>Well-established <em>in vitro</em> or <em>in vivo</em> functional studies
supportive of a damaging effect on the gene or gene product.</p>
</dd>
<dt id="term-ps4">PS4</dt><dd><p>The prevalence of the variant in affected individuals is
significantly increased compared with the prevalence in controls.</p>
</dd>
<dt id="term-pm1">PM1</dt><dd><p>Located in a mutational hot spot and/or critical and well-established
functional domain (e.g., active site of an enzyme) without benign
variation.</p>
</dd>
<dt id="term-pm2">PM2</dt><dd><p>Absent from controls (or at extremely low frequency if recessive) in
Exome Sequencing Project, 1000 Genomes Project, or Exome Aggregation
Consortium.</p>
</dd>
<dt id="term-pm3">PM3</dt><dd><p>For recessive disorders, detected in trans with a pathogenic variant.</p>
</dd>
<dt id="term-pm4">PM4</dt><dd><p>Protein length changes as a result of inframe deletions/insertions
in a non-repeat region or stop-loss variants.</p>
</dd>
<dt id="term-pm5">PM5</dt><dd><p>Novel missense change at an amino acid residue where a different
missense change determined to be pathogenic has been seen before.</p>
</dd>
<dt id="term-pm6">PM6</dt><dd><p>Assumed <em>de novo</em>, but without confirmation of paternity and maternity.</p>
</dd>
<dt id="term-pp1">PP1</dt><dd><p>Co-segregation with disease in multiple affected family members in a
gene definitively known to cause the disease</p>
</dd>
<dt id="term-pp2">PP2</dt><dd><p>Missense variant in a gene that has a low rate of benign missense
variation and in which missense variants are a common mechanism of
disease.</p>
</dd>
<dt id="term-pp3">PP3</dt><dd><p>Multiple lines of computational evidence support a deleterious effect
on the gene or gene product (conservation, evolutionary, splicing
impact, etc.).</p>
</dd>
<dt id="term-pp4">PP4</dt><dd><p>Patient’s phenotype or family history is highly specific for a
disease with a single genetic etiology.</p>
</dd>
<dt id="term-pp5">PP5</dt><dd><p>Reputable source recently reports variant as pathogenic, but the
evidence is not available to the laboratory to perform an independent
evaluation.</p>
</dd>
<dt id="term-bp1">BP1</dt><dd><p>Missense variant in a gene for which primarily truncating variants
are known to cause disease.</p>
</dd>
<dt id="term-bp2">BP2</dt><dd><p>Observed in trans with a pathogenic variant for a fully penetrant
dominant gene/disorder or observed in cis with a pathogenic variant
in any inheritance pattern.</p>
</dd>
<dt id="term-bp3">BP3</dt><dd><p>In-frame deletions/insertions in a repetitive region without a known
function.</p>
</dd>
<dt id="term-bp4">BP4</dt><dd><p>Multiple lines of computational evidence suggest no impact on gene or
gene product (conservation, evolutionary, splicing impact, etc.).</p>
</dd>
<dt id="term-bp5">BP5</dt><dd><p>Variant found in a case with an alternate molecular basis for disease.</p>
</dd>
<dt id="term-bp6">BP6</dt><dd><p>Reputable source recently reports variant as benign, but the evidence
is not available to the laboratory to perform an independent
evaluation.</p>
</dd>
<dt id="term-bp7">BP7</dt><dd><p>A synonymous (silent) variant for which splicing prediction
algorithms predict no impact to the splice consensus sequence nor the
creation of a new splice site AND the nucleotide is not highly
conserved.</p>
</dd>
<dt id="term-bs1">BS1</dt><dd><p>Allele frequency is greater than expected for disorder.</p>
</dd>
<dt id="term-bs2">BS2</dt><dd><p>Observed in a healthy adult individual for a recessive (homozygous),
dominant (heterozygous), or X-linked (hemizygous) disorder, with full
penetrance expected at an early age.</p>
</dd>
<dt id="term-bs3">BS3</dt><dd><p>Well-established in vitro or in vivo functional studies show no
damaging effect on protein function or splicing.</p>
</dd>
<dt id="term-bs4">BS4</dt><dd><p>Lack of segregation in affected members of a family.</p>
</dd>
<dt id="term-ba1">BA1</dt><dd><p>Allele frequency is &gt;5% in Exome Sequencing Project, 1000 Genomes
Project, or Exome Aggregation Consortium.</p>
</dd>
</dl>
</div>
<div class="section" id="references">
<h2>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h2>
<dl class="footnote brackets">
<dt class="label" id="acmg2015"><span class="brackets"><a class="fn-backref" href="#id1">1</a></span></dt>
<dd><p>Richards, S. et al. (2015). “Standards and guidelines for the
interpretation of sequence variants: a joint consensus recommendation of the
American College of Medical Genetics and Genomics and the Association for
Molecular Pathology”. Genetics in Medicine 17, 405–423.
DOI: 10.1038/gim.2015.30</p>
</dd>
</dl>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">

<h1 class="logo"><a href="index.html">CharGer</a></h1>



<p class="blurb">Characterization of Germline variants</p>
<h3><a href="contents.html">Table of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="index.html">Home</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="prepare_inputs.html">Preparing the Inputs</a></li>
<li class="toctree-l1"><a class="reference internal" href="cli.html">Command line interface</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">ACMG guideline</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-definitions-and-criteria">Module definitions and criteria</a></li>
<li class="toctree-l2"><a class="reference internal" href="#references">References</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="development.html">Development Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="internals/index.html">CharGer Internals</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
</ul>


<hr />
<ul>
    
    <li class="toctree-l1"><a href="https://pypi.org/project/CharGer">PyPI releases</a></li>
    
    <li class="toctree-l1"><a href="https://github.com/ding-lab/CharGer">Source Code</a></li>
    
</ul>
<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="contents.html">Documentation overview</a><ul>
      <li>Previous: <a href="cli.html" title="previous chapter">Command line interface</a></li>
      <li>Next: <a href="development.html" title="next chapter">Development Guide</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
    <div class="searchformwrapper">
        <form class="search" action="search.html" method="get">
            <input type="text" name="q" aria-labelledby="searchlabel" placeholder="Search" />
            <input type="submit" value="Go" />
        </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2020, Ding Lab.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.4.4</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
    </div>

    

    
  </body>
</html>