ACMG guideline
==============
The table and the module definitions are adapted from the original ACMG
guideline publication\ [#ACMG2015]_.


Overview
--------

+------------------+--------+------------+------------+----------+--------+--------+
|                  |        |            |            |          |        | Very   |
| ACMG modules     | Strong | Supporting | Supporting | Moderate | Strong | Strong |
+==================+========+============+============+==========+========+========+
| Population data  | |BA1|, |            |            | |PM2|    | |PS4|  |        |
|                  | |BS1|, |            |            |          |        |        |
|                  | |BS2|  |            |            |          |        |        |
+------------------+--------+------------+------------+----------+--------+--------+
| Computational    |        | |BP1|,     | |PP3|      | |PM4|,   | |PS1|  | |PVS1| |
| and              |        | |BP4|,     |            | |PM5|    |        |        |
| predictive       |        | |BP4|,     |            |          |        |        |
| data             |        | |BP7|,     |            |          |        |        |
+------------------+--------+------------+------------+----------+--------+--------+
| Functional data  | |BS3|  |            | |PP2|      | |PM1|    | |PS3|  |        |
+------------------+--------+------------+------------+----------+--------+--------+
| Segregation data | |BS4|  |            | |PP1|                          |        |
+------------------+--------+------------+------------+----------+--------+--------+
| De novo data     |        |            |            | |PM6|    | |PS2|  |        |
+------------------+--------+------------+------------+----------+--------+--------+
| Allele data      |        | |BP2|      |            | |PM3|    |        |        |
+------------------+--------+------------+------------+----------+--------+--------+
| Others           |        | |BP5|,     | |PP5|,     |          |        |        |
|                  |        | |BP6|      | |PP4|      |          |        |        |
+------------------+--------+------------+------------+----------+--------+--------+

.. |PVS1| replace:: :term:`PVS1`

.. |PS1| replace:: :term:`PS1`
.. |PS2| replace:: :term:`PS2`
.. |PS3| replace:: :term:`PS3`
.. |PS4| replace:: :term:`PS4`

.. |PM1| replace:: :term:`PM1`
.. |PM2| replace:: :term:`PM2`
.. |PM3| replace:: :term:`PM3`
.. |PM4| replace:: :term:`PM4`
.. |PM5| replace:: :term:`PM5`
.. |PM6| replace:: :term:`PM6`

.. |PP1| replace:: :term:`PP1`
.. |PP2| replace:: :term:`PP2`
.. |PP3| replace:: :term:`PP3`
.. |PP4| replace:: :term:`PP4`
.. |PP5| replace:: :term:`PP5`

.. |BP1| replace:: :term:`BP1`
.. |BP2| replace:: :term:`BP2`
.. |BP3| replace:: :term:`BP3`
.. |BP4| replace:: :term:`BP4`
.. |BP5| replace:: :term:`BP5`
.. |BP6| replace:: :term:`BP6`
.. |BP7| replace:: :term:`BP7`

.. |BS1| replace:: :term:`BS1`
.. |BS2| replace:: :term:`BS2`
.. |BS3| replace:: :term:`BS3`
.. |BS4| replace:: :term:`BS4`

.. |BA1| replace:: :term:`BA1`


Module definitions and criteria
-------------------------------

.. glossary::

    PVS1
        Null variant (nonsense, frameshift, canonical ±1 or 2 splice sites,
        initiation codon, single or multi-exon deletion) in a gene where LOF
        is a known mechanism of disease.

    PS1
        Same amino acid change as a previously established pathogenic variant
        regardless of nucleotide change.

    PS2
        *De novo* (both maternity and paternity confirmed) in a patient with
        the disease and no family history.

    PS3
        Well-established *in vitro* or *in vivo* functional studies
        supportive of a damaging effect on the gene or gene product.

    PS4
        The prevalence of the variant in affected individuals is
        significantly increased compared with the prevalence in controls.

    PM1
        Located in a mutational hot spot and/or critical and well-established
        functional domain (e.g., active site of an enzyme) without benign
        variation.

    PM2
        Absent from controls (or at extremely low frequency if recessive) in
        Exome Sequencing Project, 1000 Genomes Project, or Exome Aggregation
        Consortium.

    PM3
        For recessive disorders, detected in trans with a pathogenic variant.

    PM4
        Protein length changes as a result of inframe deletions/insertions
        in a non-repeat region or stop-loss variants.

    PM5
        Novel missense change at an amino acid residue where a different
        missense change determined to be pathogenic has been seen before.

    PM6
        Assumed *de novo*, but without confirmation of paternity and maternity.

    PP1
        Co-segregation with disease in multiple affected family members in a
        gene definitively known to cause the disease

    PP2
        Missense variant in a gene that has a low rate of benign missense
        variation and in which missense variants are a common mechanism of
        disease.

    PP3
        Multiple lines of computational evidence support a deleterious effect
        on the gene or gene product (conservation, evolutionary, splicing
        impact, etc.).

    PP4
        Patient’s phenotype or family history is highly specific for a
        disease with a single genetic etiology.

    PP5
        Reputable source recently reports variant as pathogenic, but the
        evidence is not available to the laboratory to perform an independent
        evaluation.

    BP1
        Missense variant in a gene for which primarily truncating variants
        are known to cause disease.

    BP2
        Observed in trans with a pathogenic variant for a fully penetrant
        dominant gene/disorder or observed in cis with a pathogenic variant
        in any inheritance pattern.

    BP3
        In-frame deletions/insertions in a repetitive region without a known
        function.

    BP4
        Multiple lines of computational evidence suggest no impact on gene or
        gene product (conservation, evolutionary, splicing impact, etc.).

    BP5
        Variant found in a case with an alternate molecular basis for disease.

    BP6
        Reputable source recently reports variant as benign, but the evidence
        is not available to the laboratory to perform an independent
        evaluation.

    BP7
        A synonymous (silent) variant for which splicing prediction
        algorithms predict no impact to the splice consensus sequence nor the
        creation of a new splice site AND the nucleotide is not highly
        conserved.

    BS1
        Allele frequency is greater than expected for disorder.

    BS2
        Observed in a healthy adult individual for a recessive (homozygous),
        dominant (heterozygous), or X-linked (hemizygous) disorder, with full
        penetrance expected at an early age.

    BS3
        Well-established in vitro or in vivo functional studies show no
        damaging effect on protein function or splicing.

    BS4
        Lack of segregation in affected members of a family.

    BA1
        Allele frequency is >5% in Exome Sequencing Project, 1000 Genomes
        Project, or Exome Aggregation Consortium.


References
----------

.. [#ACMG2015] Richards, S. et al. (2015). "Standards and guidelines for the
   interpretation of sequence variants: a joint consensus recommendation of the
   American College of Medical Genetics and Genomics and the Association for
   Molecular Pathology". Genetics in Medicine 17, 405–423.
   DOI: 10.1038/gim.2015.30
